Mainnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs

```diff
@@ -0,0 +1,639 @@
+        ClearApproval(tokenId);
+
+        RemoveNFToken(from, tokenId);
+        AddNFToken(to, tokenId);
+
+        LogTransfer(from, to, tokenId);
+    }
+
+    /// <summary>
+    /// Removes a NFT from owner.
+    /// </summary>
+    /// <remarks>Use and override this function with caution. Wrong usage can have serious consequences.</remarks>
+    /// <param name="from">Address from wich we want to remove the NFT.</param>
+    /// <param name="tokenId">Which NFT we want to remove.</param>
+    private void RemoveNFToken(Address from, ulong tokenId)
+    {
+        Assert(GetIdToOwner(tokenId) == from);
+        var tokenCount = GetOwnerToNFTokenCount(from);
+        SetOwnerToNFTokenCount(from, checked(tokenCount - 1));
+        this.PersistentState.Clear(GetIdToOwnerKey(tokenId));
+
+        ulong index = GetIndexOfOwnerByToken(from, tokenId);
+        ulong lastIndex = tokenCount - 1;
+
+        if (index != lastIndex)
+        {
+            ulong lastToken = GetTokenOfOwnerByIndex(from, lastIndex);
+            SetIndexOfOwnerByToken(from, lastToken, index);
+            SetTokenOfOwnerByIndex(from, index, lastToken);
+        }
+
+        ClearTokenOfOwnerByIndex(from, lastIndex);
+        ClearIndexOfOwnerByToken(from, tokenId);
+    }
+
+    /// <summary>
+    /// Assignes a new NFT to owner.
+    /// </summary>
+    /// <remarks>Use and override this function with caution. Wrong usage can have serious consequences.</remarks>
+    /// <param name="to">Address to which we want to add the NFT.</param>
+    /// <param name="tokenId">Which NFT we want to add.</param>
+    private void AddNFToken(Address to, ulong tokenId)
+    {
+        Assert(GetIdToOwner(tokenId) == Address.Zero);
+
+        SetIdToOwner(tokenId, to);
+        ulong currentTokenAmount = GetOwnerToNFTokenCount(to);
+        SetOwnerToNFTokenCount(to, checked(currentTokenAmount + 1));
+
+        var index = currentTokenAmount;
+        SetIndexOfOwnerByToken(to, tokenId, index);
+        SetTokenOfOwnerByIndex(to, index, tokenId);
+    }
+
+    /// <summary>
+    /// Actually perform the safeTransferFrom.
+    /// </summary>
+    /// <param name="from">The current owner of the NFT.</param>
+    /// <param name="to">The new owner.</param>
+    /// <param name="tokenId">The NFT to transfer.</param>
+    /// <param name="data">Additional data with no specified format, sent in call to 'to' if it is a contract.</param>
+    private void SafeTransferFromInternal(Address from, Address to, ulong tokenId, byte[] data)
+    {
+        CanTransfer(tokenId);
+        ValidNFToken(tokenId);
+
+        Address tokenOwner = GetIdToOwner(tokenId);
+        Assert(tokenOwner == from);
+        EnsureAddressIsNotEmpty(to);
+
+        TransferInternal(to, tokenId);
+
+        if (this.PersistentState.IsContract(to))
+        {
+            ITransferResult result = this.Call(to, 0, "OnNonFungibleTokenReceived", new object[] { this.Message.Sender, from, tokenId, data }, 0);
```

Comment:
This method `OnNonFungibleTokenReceived` is in interfaces but is not implemented anywhere, where is this call supposed to be going to?

Reply:
The contract which wants to be notified when they get a token(nft) should to implement that interface actually.  If NFT sends to a contract address then NFT contract notifies the client contract by doing external call.